src/lib-imap/imap-capability-list.h

```diff
@@ -0,0 +1,81 @@
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "lib.h"
+#include "array.h"
+
+/* bitflags to represent different conditions of
+   the current imap session */
+
+#define CAP_NEVER       0        /* never show this capability */
+#define CAP_PREAUTH     (1 << 0) /* show this cap before login */
+#define CAP_POSTAUTH    (1 << 1) /* show this cap after login */
+#define CAP_TLSACTIVE   (1 << 2) /* show this cap if tls is active */
+#define CAP_TLSINACTIVE (1 << 3) /* show this cap if tls is inactive */
+#define CAP_SSLACTIVE   (1 << 4) /* show this cap if ssl is active */
+#define CAP_SSLINACTIVE (1 << 5) /* show this cap if ssl is inactive */
+#define CAP_NO_LOGIN    (1 << 6) /* show this cap if login is disabled */
+/* NOTE: connections from localhost are 'secure' regardless of SSL/TLS! */
+#define CAP_SECURE      (1 << 7) /* show this cap if we are secured */
+#define CAP_INSECURE    (1 << 8) /* show this cap if we aren't secured */
+#define CAP_ALWAYS      0x7FFFFFFF /* always show this cap */
+
+/* CAP_REQUIRE_ALL indicates that the rest of the conditions
+   must be met *exactly* to show this capability, otherwise
+   if *at least* 1 condition is met the capability will be
+   shown */
+#define CAP_REQUIRE_ALL (1 << 31)
+
+/*  the initial size of the capability list
+    it gets automatically increased but this
+    is the size that we start at */
+#define INITIAL_CAP_LIST_SIZE 5
+/*  The amount to grow the capability list
+    each time it fills up */
+#define CAP_LIST_GROW_AMOUNT  1
+
+/* XXX: This really should be integrated with the imap_commands */
+struct imap_capability {
+    char *command;
+    uint32_t flags;
+};
+
+/* define our array type for the imap capability array */
+ARRAY_DEFINE_TYPE(imap_capability_array, struct imap_capability);
+
+/* define the structure that holds the capability pool and command array */
+struct imap_capability_list {
+    /* reference count */
+    int refcount;
+    /* imap capability list memory pool */
+    pool_t pool;
+    /* array of imap_capability objects */
+    ARRAY_TYPE(imap_capability_array) imap_capabilities;
+};
+
+/* initialize the imap capability list */
+struct imap_capability_list *
+imap_capability_list_create(const char *initial_entries);
+/* increment reference to an imap capability list */
+void imap_capability_list_ref(struct imap_capability_list *capability_list);
+/* decrement reference to an imap capability list */
+void imap_capability_list_unref(struct imap_capability_list **capability_list);
+/* add an entry to the imap capability list */
+void imap_capability_list_add(struct imap_capability_list *capability_list,
+        const char *command, uint32_t flags);
+/* remove an entry from the imap capability list */
+void imap_capability_list_remove(struct imap_capability_list *capability_list,
+        const char *command);
+/* split a string by space and add each entry to the imap capability list */
+void imap_capability_list_append_string(
```

Comment:
It is ok to go over 80 columns in dovecot.  It avoids some awkward line breaks given the large number of very long symbol/identifier names.